Sources/MongoSwift/BSON/BSONDecoder.swift

```diff
@@ -314,39 +314,54 @@ internal struct _BSONDecodingStorage {
 
 /// Extend _BSONDecoder to add methods for "unboxing" values as various types.
 extension _BSONDecoder {
-    fileprivate func unboxBSONValue<T: BSONValue>(_ value: BSONValue, as type: T.Type) throws -> T {
+    fileprivate func unboxBSON<T>(_ value: BSON, f: (BSON) -> T?) throws -> T {
         // We throw in the case of BSONNull because nulls should be requested through decodeNil().
-        guard !(value is BSONNull) else {
+        guard value != .null else {
             throw DecodingError.valueNotFound(
-                type,
-                DecodingError.Context(codingPath: self.codingPath,
-                                      debugDescription: "Expected a non-null type."))
+                    T.self,
+                    DecodingError.Context(codingPath: self.codingPath,
+                                          debugDescription: "Expected a non-null type."))
+        }
+
+        guard let typed = f(value) else {
+            throw DecodingError._typeMismatch(at: self.codingPath, expectation: T.self, reality: value.bsonValue)
+        }
+        return typed
+    }
+
+    fileprivate func unboxBSONValue<T: BSONValue>(_ value: BSON, as type: T.Type) throws -> T {
```

Comment:
can you explain what the difference is between `unboxBSON` and`unboxBSONValue` now / why we need both? (possibly should go into docstrings.)